include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -1377,6 +1377,330 @@ namespace internal
 
 
 
+  /**
+   * This class chooses an appropriate evaluation strategy based on the
+   * template parameters and the shape_info variable which contains runtime
+   * parameters for the strategy underlying FEEvaluation::evaluate(), i.e.
+   * this calls internal::FEEvaluationImpl::evaluate(),
+   * internal::FEEvaluationImplCollocation::evaluate() or
+   * internal::FEEvaluationImplTransformToCollocation::evaluate() with
+   * appropriate template parameters. In case the template parameters
+   * fe_degree and n_q_points_1d contain valid information (i.e. fe_degree>-1
+   * and n_q_points_1d>0), we simply pass these values to the respective
+   * template specializations.  Otherwise, we perform a runtime matching of
+   * the runtime parameters to find the correct specialization. This matching
+   * currently supports $0\leq fe\_degree \leq 9$ and $degree+1\leq
+   * n\_q\_points\_1d\leq fe\_degree+2$.
+   */
+  template <int dim, typename Number>
+  struct FEEvaluationImplEvaluateSelector
+  {
+    template <int fe_degree, int n_q_points_1d>
+    static void
+    run(const unsigned int                                      n_components,
+        const EvaluationFlags::EvaluationFlags                  evaluation_flag,
+        const internal::MatrixFreeFunctions::ShapeInfo<Number> &shape_info,
+        Number *values_dofs_actual,
+        Number *values_quad,
+        Number *gradients_quad,
+        Number *hessians_quad,
+        Number *scratch_data);
+  };
```

Comment:
I am leaning towards using variadic template `args...` because the only way I see those could be generalized is by having a new struct that hides the function-specific part, which is also not pretty.